### PR TITLE
[No QA][TS migration] Migrate 'UserUtils.js' lib to TypeScript

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

Migration of 'UserUtils.js' to Typescript.

### Fixed Issues
$ https://github.com/Expensify/App/issues/24882
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
N/A

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/46095609/d7f7538b-0065-4cf9-b23a-bbefdcda23b9

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/46095609/b8107a67-5f54-455b-8c5c-0d5f5464c807

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/46095609/1aeb81bc-7c23-4cc0-901e-81bb467b2ebb

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/46095609/49e9abaa-1133-4b5e-9157-1156194e755e

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/46095609/00edfbc7-008c-4c41-99a9-663fec55981d

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/46095609/428fd682-7312-4fbb-8add-948f7e272e89

</details>

### CODE REVIEW

**Review by allroundexperts — APPROVED** [review]

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [inline src/libs/UserUtils.ts:108]: I this type `string` correct? if it is, why would we need to check `typeof avatarURL === 'string'` below?

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [inline src/libs/UserUtils.ts:156]: same question as above

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [inline src/libs/UserUtils.ts:32]: Is the type `loginList: LoginList` correct? if it is, we wouldn't need the question mark in `loginList?`

**Review by aldo-expensify — CHANGES_REQUESTED** [review]
Some types don't seem to be correct

File: src/libs/UserUtils.ts
Comment (blazejkustra) [reply src/libs/UserUtils.ts:156]: source is of type `React.FC<SvgProps> | string` so the condition here seems to be valid

File: src/libs/UserUtils.ts
Comment (blazejkustra) [reply src/libs/UserUtils.ts:108]: Adjusted!

File: src/libs/UserUtils.ts
Comment (blazejkustra) [reply src/libs/UserUtils.ts:32]: Yes it's correct, the thing is that this function is still used in some JS files and wrong values are passed and tests were failing. Once we get to these files that use this function the underlying issue will be fixed.

Either way, it's okay to make code safe and use `nullish coalescing` and `optional chaining` even if types says it's not necessary.

**Comment by blazejkustra** [comment]
@aldo-expensify Let's wait for Callstack review on this PR before merging.

File: src/libs/UserUtils.ts
Comment (fabioh8010) [inline src/libs/UserUtils.ts:52]: Could we extract `'' | 'error' | 'info'` into a separate type?

File: src/ONYXKEYS.ts
Comment (fabioh8010) [inline src/ONYXKEYS.ts:324]: Maybe could we change the type back to `Login`? `LoginList` looks so misleading to me, we are not handling any lists here. I understand that the onyx key is LOGIN_LIST, but if we have opportunity to at least name the data structure correctly we should leave it.

Comment (blazejkustra) [reply src/ONYXKEYS.ts]: Sure! ➕

**Review by fabioh8010 — COMMENTED** [review]
@blazejkustra LGTM, but we have a type error reported by `typecheck`.

**Comment by fabioh8010** [comment]
@blazejkustra We have another lint error 😅

**Comment by blazejkustra** [comment]
Tests are failing due to [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1695796810686659) thread.

**Review by fabioh8010 — APPROVED** [review]
LGTM, CK still need to do the checklist.

**Comment by kubabutkiewicz** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
  - [x] I verified the steps for local testing are in the `Tests` section
  - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
  - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
  - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
  - [x] Android / native
  - [x] Android / Chrome
  - [x] iOS / native
  - [x] iOS / Safari
  - [x] MacOS / Chrome / Safari
  - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
  - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
  - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
  - [x] I verified that comments were added to code that is not self explanatory
  - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
  - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
  - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
  - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
  - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
  - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
  - [x] A similar component doesn't exist in the codebase
  - [x] All props are defined accurately and each prop has a `/** comment above it */`
  - [x] The file is named correctly
  - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
  - [x] The only data being stored in the state is data necessary for rendering and nothing else
  - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
  - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
  - [x] All JSX used for rendering exists in the render method
  - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
  - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
  - [x] A similar style doesn't already exist
  - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/25230417/160f0ca4-1be4-44f9-912d-492729500d50




</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/25230417/95ee9d2e-13e0-40b5-bc8a-2dbacabb854a




</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/25230417/9f659c4f-77d8-424c-babd-53eb24c71d76




</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/25230417/6aa627d8-6639-40fc-9df0-6e056c3a3b8a




</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/25230417/452d6e7c-c4d8-4804-a9ce-2123b33e950f




</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/25230417/f65e30fe-4c97-4fab-8f1d-bf96d84801fc




</details>

**Review by kubabutkiewicz — APPROVED** [review]

**Comment by blazejkustra** [comment]
@aldo-expensify PR is all yours now!

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [reply src/libs/UserUtils.ts:156]: True, I think I didn't read it correctly at the time.

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [reply src/libs/UserUtils.ts:35]: > Either way, it's okay to make code safe and use nullish coalescing and optional chaining even if types says it's not necessary.

I disagree with this, I expect to be able to trust our types, but I understand that this may be acceptable while in a transitional period

I'm not sure why we didn't prefer `loginList: Login | undefined | null` if we had cases where we are passing `undefined` or `null` 🤷

**Review by aldo-expensify — APPROVED** [review]
Approving and merging because I think this is mostly correct, if we want to do something regarding [this](https://github.com/Expensify/App/pull/27778/files#r1332355479) we can do it in a follow up!

Thanks for the work!

File: src/libs/UserUtils.ts
Comment (aldo-expensify) [inline src/libs/UserUtils.ts:46]: This change caused this regression: https://github.com/Expensify/App/issues/28643

**Pull request merged**
